libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts

```diff
@@ -1038,165 +1027,314 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
+                    // Property and value match ambiguous entity.
+                    assignment.operation = AdaptiveEvents.chooseEntity;
+                    assignment.isExpected = true;
+                } else {
+                    // Assign default operator.
+                    assignment.operation = this.defaultOperation(assignment, askDefault, dialogDefault);
+                }
+            }
         }
 
-        // Unassigned entities
-        const entityPreferences = this.entityPreferences(null);
-        for (const key in entities) {
-            if (!usedEntityType.has(key) && entityPreferences.indexOf(key) >= 0) {
-                for (const entity of entities[key]) {
-                    if (!usedEntity.has(entity.name)) {
-                        candidates.push({
-                            entity: entity,
-                            operation: entity.operation,
-                            property: entity.property,
-                        });
+        // Add choose property matches.
+        if (lastEvent === AdaptiveEvents.chooseProperty) {
+            for (const alternatives of Object.values(entities)) {
+                alternatives.forEach((alternative) => {
+                    if (!alternative.value) {
+                        // If alternative matches one alternative, it answers chooseProperty.
+                        const matches = nextAssignment.alternatives.filter((a) =>
+                            this.matchesAssignment(alternative, a)
+                        );
+                        if (matches.length === 1) {
+                            assignments.push(
+                                new EntityAssignment({
+                                    value: alternative,
+                                    operation: AdaptiveEvents.chooseProperty,
+                                    isExpected: true,
+                                })
+                            );
+                        }
                     }
+                });
+            }
+        }
+
+        // Add pure operations.
+        for (const alternatives of Object.values(entities)) {
+            alternatives.forEach((alternative) => {
+                if (alternative.operation && !alternative.property && !alternative.value) {
+                    assignments.push(
+                        new EntityAssignment({
+                            value: alternative,
+                            property: null,
+                            operation: alternative.operation,
+                            isExpected: false,
+                        })
+                    );
                 }
+            });
+        }
+
+        // Preserve expectedProperties if there is no property.
+        for (const assignment of assignments) {
+            if (!assignment.property) {
+                assignment.expectedProperties = expected;
             }
         }
 
-        return candidates;
+        return assignments;
     }
 
     /**
```

Comment:
You could strip these `@private` JSDoc comments, they're not required anymore.

Reply:
Hallelujah!